pkg/kubecost/asset.go

```diff
@@ -2565,24 +2645,63 @@ func (as *AssetSet) Clone() *AssetSet {
 		warnings = nil
 	}
 
-	return &AssetSet{
-		Window:      NewWindow(&s, &e),
-		aggregateBy: aggregateBy,
-		assets:      assets,
-		Errors:      errors,
-		Warnings:    warnings,
+	var anyMap map[string]*Any
+	if as.Any != nil {
+		anyMap = make(map[string]*Any, len(as.Any))
+	}
+	var cloudMap map[string]*Cloud
+	if as.Cloud != nil {
+		cloudMap = make(map[string]*Cloud, len(as.Cloud))
+	}
+	var clusterManagementMap map[string]*ClusterManagement
+	if as.ClusterManagement != nil {
+		clusterManagementMap = make(map[string]*ClusterManagement, len(as.ClusterManagement))
+	}
+	var disksMap map[string]*Disk
+	if as.Disks != nil {
+		disksMap = make(map[string]*Disk, len(as.Disks))
+	}
+	var networkMap map[string]*Network
+	if as.Network != nil {
+		networkMap = make(map[string]*Network, len(as.Network))
+	}
+	var nodesMap map[string]*Node
+	if as.Nodes != nil {
+		nodesMap = make(map[string]*Node, len(as.Nodes))
+	}
+	var loadBalancersMap map[string]*LoadBalancer
+	if as.LoadBalancers != nil {
+		loadBalancersMap = make(map[string]*LoadBalancer, len(as.LoadBalancers))
 	}
-}
 
-// Each invokes the given function for each Asset in the set
-func (as *AssetSet) Each(f func(string, Asset)) {
-	if as == nil {
-		return
+	var sharedAssetsMap map[string]*SharedAsset
+	if as.SharedAssets != nil {
+		sharedAssetsMap = make(map[string]*SharedAsset, len(as.SharedAssets))
 	}
 
-	for k, a := range as.assets {
-		f(k, a)
+	assetSet := &AssetSet{
+		Window:            NewWindow(&s, &e),
+		AggregationKeys:   aggregateBy,
+		Assets:            make(map[string]Asset, len(as.Assets)),
+		Any:               anyMap,
+		Cloud:             cloudMap,
+		ClusterManagement: clusterManagementMap,
+		Disks:             disksMap,
+		Network:           networkMap,
+		Nodes:             nodesMap,
+		LoadBalancers:     loadBalancersMap,
+		SharedAssets:      sharedAssetsMap,
+		Errors:            errors,
+		Warnings:          warnings,
 	}
+
+	for k, v := range as.Assets {
+		newAsset := v.Clone()
+		assetSet.Assets[k] = newAsset
+		addToConcreteMap(assetSet, k, newAsset)
+	}
```

Comment:
Should we just use Insert() here?

Reply:
I looked closely at this, and I don't think `Insert()` is necessary because we are cloning directly from another AssetSet, so we won't ever have the case where a new Asset should be added to an existing asset, or we will ever have to expand the window.

Although, I think I just realized something wrong with Insert() while looking into this a bit more.